updates `Results.summary` method to include `form` arg

Burhan-Q opened the pull request:
- Include arg `form` to select either `xyxy` or `xywh` formats returned for bounding box coordinates. 
- Also looks for `"n"` in `form` argument to override `normalize` argument
- Checked with `detect`, `segment`, and `pose` tasks and all worked well

<!--
Thank you 🙏 for your contribution to [Ultralytics](https://ultralytics.com) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. **Check for Existing Contributions**: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. **Link Related Issues**: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. **Elaborate Your Changes**: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. **Ultralytics Contributor License Agreement (CLA)**: To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    _I have read the CLA Document and I hereby sign the CLA_

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing). Your adherence to these guidelines ensures a faster and more effective review process.
--->


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced summary output for YOLO model predictions with format flexibility 🚀

### 📊 Key Changes
- Added `test_results_summary()` in tests to rigorously check model predictions' summary output.
- Updated the `summary()` method in `results.py` to support different box formats (`xyxy`, `xywh`) and normalization directly in the summary output.
- Introduced additional arguments in the `summary()` and `tojson()` methods for specifying the format of the box (e.g., `format="xyxy"`).

### 🎯 Purpose & Impact
- **Purpose**: To make model predictions' summaries more versatile and useful by adding flexibility in the output format. This is particularly useful for various post-processing and visualization needs where different box formats and normalized coordinates might be required.
- **Impact**: Users can now receive predictions summaries in their desired box format directly, eliminating the need for manual post-processing. This enhances the usability of the Ultralytics framework for both development and operational tasks, making it easier to integrate into various applications and workflows. 🌐

🔍 By providing detailed summaries in a flexible format, this update empowers developers to streamline their workflows and fosters more creative uses of detection models, potentially leading to innovative applications in fields like surveillance, autonomous driving, and retail analytics.

Burhan-Q committed:
updates `Results.summary` method to include `form` arg

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Burhan-Q committed:
adds in-depth test for `Results.summary` method

Burhan-Q committed:
fix type annotations

Burhan-Q commented:
Added test for `Results.summary()` PR ready for review

glenn-jocher commented:
@Burhan-Q hey thanks, looks good! Can you update the arg name to 'format'?

There's a few TODO here (not for this PR but in general on the results summary):

- Add OBB
- Add Classify
- Investigate normalization implementation as I feel most of these formats already include normalized values i.e. `xyxyn` so we may be normalizing twice.

Burhan-Q committed:
updated `form` argument to `format` as requested in PR comment, `test_results_summary` passes locally

Burhan-Q committed:
reuses normalized attributes assures rounding sticks after conversion to list

Burhan-Q commented:
@glenn-jocher updated to `format` as requested. Additionally, changed to reuse the existing object attributes for normalization instead of recalculating.

I also found that rounding wasn't holding after converting to `list`, which I tested with:

```py
from ultralytics import YOLO

model = YOLO("yolov8n-pose.pt")
results = model.predict("ultralytics/assets/bus.jpg")

x, y, visible = results[0].keypoints[0].data[0].cpu().unbind(dim=1)
x.numpy().round(5).tolist()
>>> [143.7509307861328, 149.05174255371094, 133.68670654296875, 0.0, 107.28692626953125, 163.1969757080078, 88.74665832519531, 191.32513427734375, 115.7267837524414, 157.353759765625, 149.51995849609375, 156.18040466308594, 98.22109985351562, 182.1588134765625, 90.46875762939453, 189.28085327148438, 72.58592987060547]

x.numpy().astype(float).round(5).tolist()
>>> [143.75093, 149.05174, 133.68671, 0.0, 107.28693, 163.19698, 88.74666, 191.32513, 115.72678, 157.35376, 149.51996, 156.1804, 98.2211, 182.15883, 90.46876, 189.28085, 72.58593]
```
which is why I added the `astype(float)` method as well.

Burhan-Q committed:
Merge branch 'main' into summary_update

glenn-jocher committed:
Merge branch 'main' into summary_update

glenn-jocher committed:
Merge branch 'main' into summary_update

glenn-jocher committed:
Merge branch 'main' into summary_update

glenn-jocher committed:
Merge branch 'main' into summary_update

glenn-jocher committed:
Merge branch 'main' into summary_update

glenn-jocher committed:
Merge branch 'main' into summary_update

glenn-jocher committed:
Merge branch 'main' into summary_update

Burhan-Q committed:
Merge branch 'main' into summary_update

glenn-jocher committed:
Merge branch 'main' into summary_update

glenn-jocher committed:
Merge branch 'main' into summary_update

glenn-jocher committed:
Merge branch 'main' into summary_update

Burhan-Q committed:
Merge branch 'main' into summary_update

glenn-jocher committed:
Merge branch 'main' into summary_update

Burhan-Q committed:
Merge branch 'main' into summary_update

UltralyticsAssistant committed:
Merge 'main' into PR #9307 by https://ultralytics.com/actions

UltralyticsAssistant committed:
Merge branch 'main' into summary_update

UltralyticsAssistant committed:
Merge branch 'main' into summary_update

Burhan-Q committed:
Merge branch 'main' into summary_update

UltralyticsAssistant committed:
Merge branch 'main' into summary_update

UltralyticsAssistant committed:
Merge branch 'main' into summary_update

Burhan-Q committed:
Merge branch 'main' into summary_update

Burhan-Q committed:
Merge branch 'main' into summary_update

Burhan-Q committed:
Merge branch 'main' into summary_update

Burhan-Q committed:
Merge branch 'main' into summary_update

Burhan-Q committed:
Merge branch 'main' into summary_update

Burhan-Q committed:
Merge branch 'main' into summary_update

glenn-jocher committed:
Merge branch 'main' into summary_update
